compiler/src/main/kotlin/io/github/apl_cornell/viaduct/analysis/NameAnalysis.kt

```diff
@@ -480,48 +518,132 @@
+                    rhs.check(hosts, labelVariables, sourceLocation)
+                }
+
+                is LabelOr -> {
+                    lhs.check(hosts, labelVariables, sourceLocation)
+                    rhs.check(hosts, labelVariables, sourceLocation)
+                }
+
+                else -> {}
+            }
+        }
+
+        /** check if a [LabelNode] has undeclared [Host] or [LabelVariable] */
+        fun LabelNode.check(node: Node) {
+            val hosts = node.hostDeclarations.keys
+            val labelVariables = node.labelVariables()
+            value.check(hosts, labelVariables, sourceLocation)
+        }
+
         fun check(node: Node) {
             // Check that name references are valid
             when (node) {
                 is ParameterNode ->
                     node.protocol?.check()
+
                 is ReadNode ->
                     declaration(node)
+
                 is QueryNode ->
                     declaration(node)
+
                 is LetNode ->
                     node.protocol?.check()
+
                 is DeclarationNode ->
                     node.protocol?.check()
+
                 is UpdateNode ->
                     declaration(node)
+
                 is OutParameterInitializationNode ->
                     declaration(node)
+
                 is ObjectReferenceArgumentNode ->
                     declaration(node)
+
                 is OutParameterArgumentNode ->
                     declaration(node)
+
                 is FunctionCallNode ->
                     declaration(node)
+
                 is BreakNode ->
                     correspondingLoop(node)
+
                 is CommunicationNode ->
                     declaration(node)
             }
             // Check that there are no name clashes
             when (node) {
                 is LetNode ->
                     node.temporaryDefinitions.put(node.name, node)
+
                 is DeclarationNode ->
                     node.objectDeclarations.put(node.name, node)
+
                 is InfiniteLoopNode ->
                     node.jumpTargets.put(node.jumpLabel, node)
+
                 is ProgramNode -> {
                     // Forcing these thunks
                     node.hostDeclarations
                     node.functionDeclarations
                 }
             }
+            // Check that LabelVariables and Hosts area declared
+            when (node) {
+                is DeclarationNode ->
+                    node.objectType.labelArguments?.first()?.check(node)
+
+                is DeclassificationNode -> {
+                    node.fromLabel?.check(node)
+                    node.toLabel.check(node)
+                }
+
+                is EndorsementNode -> {
+                    node.fromLabel.check(node)
+                    node.toLabel?.check(node)
+                }
+
+                is ParameterNode -> {
+                    node.objectType.labelArguments?.first()?.check(node)
```

Comment:
Check all arguments not just the first. You can use `.forEach` instead of `.first()`.